xray/services/scan.go

```diff
@@ -0,0 +1,205 @@
+	requestBody, err := json.Marshal(scanParams.Graph)
+	if err != nil {
+		return "", errorutils.CheckError(err)
+	}
+	url := ss.XrayDetails.GetUrl() + scanGraphAPI
+	if scanParams.ProjectKey != "" {
+		url += projectQueryParam + scanParams.ProjectKey
+	} else if scanParams.RepoPath != "" {
+		url += repoPathQueryParam + scanParams.RepoPath
+	}
+	resp, body, err := ss.client.SendPost(url, requestBody, &httpClientsDetails)
+	if err != nil {
+		return "", err
+	}
+
+	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated {
+		return "", errorutils.CheckError(errors.New("Server Response: " + resp.Status))
+	}
+	scanResponse := RequestScanResponse{}
+	json.Unmarshal(body, &scanResponse)
+	return scanResponse.ScanId, err
+}
+
+func (ss *ScanService) GetScanGraphResults(scanId string) (*ScanResponse, error) {
+	maxWaitMinutes := defaultMaxWaitMinutes
+	if ss.MaxWaitMinutes > 0 {
+		maxWaitMinutes = ss.MaxWaitMinutes
+	}
+	httpClientsDetails := ss.XrayDetails.CreateHttpClientDetails()
+	utils.SetContentType("application/json", &httpClientsDetails.Headers)
+
+	message := fmt.Sprintf("Sync: Get Scan Graph Results. Scan ID:%s...", scanId)
+	// The scan request may take time to be calculated, we expect to receive a 202 response until the compleation of the requested scan.
+	ticker := time.NewTicker(defaultSyncSleepInterval)
+	timeout := make(chan bool)
+	errChan := make(chan error)
+	resultChan := make(chan []byte)
+	endPoint := ss.XrayDetails.GetUrl() + scanGraphAPI + "/" + scanId
+	go func() {
+		for {
+			select {
+			case <-timeout:
+				errChan <- errorutils.CheckError(errors.New("Timeout for sync get scan graph results."))
+				resultChan <- nil
+				return
+			case _ = <-ticker.C:
+				log.Debug(message)
+				resp, body, _, err := ss.client.SendGet(endPoint, true, &httpClientsDetails)
+				if err != nil {
+					errChan <- err
+					resultChan <- nil
+					return
+				}
+				if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusAccepted {
+					errChan <- errorutils.CheckError(errors.New("Server Response: " + resp.Status))
+					resultChan <- nil
+					return
+				}
+				// Got the full valid response.
+				if resp.StatusCode == http.StatusOK {
+					errChan <- nil
+					resultChan <- body
+					return
+				}
+			}
+		}
+	}()
+	// Make sure we don't wait forever
+	go func() {
+		time.Sleep(maxWaitMinutes)
+		timeout <- true
+	}()
+	// Wait for result or error
+	err := <-errChan
+	body := <-resultChan
+	ticker.Stop()
+	if err != nil {
+		return nil, err
+	}
+	scanResponse := ScanResponse{}
+	json.Unmarshal(body, &scanResponse)
+	return &scanResponse, err
+}
+
+type XrayGraphScanParams struct {
+	// A path in Artifactory that this Artifact is intended to be deployed to.
+	// This will provide a way to extract the watches that should be applied on this graph
+	RepoPath   string
+	ProjectKey string
+	Graph      *GraphNode
+}
+
+type GraphNode struct {
+	// Component Id in the JFrog standard.
```

Comment:
Let's describe what is the component Id. For example, we can mention that it can be `<groupId>:<artifactId>:<version>` or `<artifactId>:<version>`

Reply:
What do you think about adding a link?
https://www.jfrog.com/confluence/display/JFROG/Xray+REST+API#XrayRESTAPI-ComponentIdentifiers
(I'm not sure if it is a good idea...)

Reply:
Also possible